marshaler.go

```diff
@@ -643,13 +643,14 @@ func (enc *Encoder) encodeTableInline(b []byte, ctx encoderCtx, t table) ([]byte
 	return b, nil
 }
 
-var (
-	errNilInterface   = errors.New("nil interface not supported")
-	textMarshalerType = reflect.TypeOf(new(encoding.TextMarshaler)).Elem()
-)
+var errNilInterface = errors.New("nil interface not supported")
+
+func textMarshalerType() reflect.Type {
+	return reflect.TypeOf(new(encoding.TextMarshaler)).Elem()
+}
 
```

Comment:
I usually try to avoid global vars as much as possible and replace it with a function.
I'm pretty sure the compiler will inline the function and replace the call anyway and it stops from having a global var.

The only exception I usually do is for errors (no choice) and for compiler updated vars (like a version, for ex).

The benefit is to be 100% sure this var will never be overwritten
The drawback is the hackiness of it

I know this can be controversial, so debate is open!
I can bench as well if you re not convinced there is no perf cost.

Reply:
Benchmarking with the following code, the difference is significant:

```go
func BenchmarkTextMarshalerFn(b *testing.B) {
	v := reflect.ValueOf(struct{}{})
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		v.Type().Implements(textMarshalerType())
	}
}

var textMarshalerTypeVar = reflect.TypeOf(new(encoding.TextMarshaler)).Elem()

func BenchmarkTextMarshalerVar(b *testing.B) {
	v := reflect.ValueOf(struct{}{})
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		v.Type().Implements(textMarshalerTypeVar)
	}
}
```

```
name                 time/op
TextMarshalerFn-32   53.9ns ± 2%
TextMarshalerVar-32  6.87ns ± 2%

name                 alloc/op
TextMarshalerFn-32    16.0B ± 0%
TextMarshalerVar-32   0.00B     

name                 allocs/op
TextMarshalerFn-32     1.00 ± 0%
TextMarshalerVar-32    0.00     
```

Given that there is an alloc every op, it does not seem like the compiler is about to optimize this out (haven't double checked the emitted instructions).

The benefit seem slim here, given that to be overwritten the value would need to be of the same type to pass compilation and would need to be close enough to pass the test suite. I wish Go would let us define those as const or have literal types, but unfortunately the next best thing is to compute that value once.

Reply:
this just me shows once again that I shouldn't trust my intuition, stop talking and just bench it :).
I did bench this change in different situations, but always ended up being the same results and no alloc. I guess this one of the benefit of participating in OSS: getting exposed to more use cases :).

I'll remove the two global var commits